### PR TITLE
add defcustom cider-use-ssh-to-infer-remote-ports

### PR DESCRIPTION
**Pull request by aiba**
This is just an idea to address #1544.  If the cider maintainers want it, I can update this PR to also include documentation of the new defcustom.

### CODE REVIEW

**Comment by bbatsov** [comment]
I don't use tramp, so I don't really care about this much. With me it's perfectly fine to just delete everything tramp-related, as it has been nothing but a pain in the ass for me. At this point I hate tramp (both because of CIDER and Projectile). :-) 

On the other hand a lot of people seem to be using it for some weird reason... In principle I'm fine with the suggested change.

File: cider.el [hunk 1]
```diff
@@ -1338,14 +1343,15 @@ of remote SSH hosts."
...
-      (let ((vec (vector "sshx" nil host "" nil))
-            ;; change dir: user might want to connect to a different remote
-            (dir (when (file-remote-p default-directory)
-                   (with-parsed-tramp-file-name default-directory cur
-                     (when (string= cur-host host) default-directory)))))
-        (tramp-maybe-open-connection (cider--tramp-file-name vec))
-        (with-current-buffer (tramp-get-connection-buffer (cider--tramp-file-name vec))
-          (cider-locate-running-nrepl-ports dir))))))
+      (when cider-use-ssh-to-infer-remote-nrepl-ports
+        (let ((vec (vector "sshx" nil host "" nil))
```
Comment (bbatsov) [inline cider.el]: I wonder if we shouldn't move the remote logic to a separate function.

File: cider.el [hunk 2]
```diff
@@ -1327,6 +1327,11 @@ Tramp version starting 26.1 is using a `cl-defstruct' rather than vanilla VEC."
       (make-tramp-file-name :method (elt vec 0)
                             :host   (elt vec 2)))))
 
+(defcustom cider-use-ssh-to-infer-remote-nrepl-ports t
```
Comment (bbatsov) [inline cider.el]: Probably we can shorten this simply to `cider-infer-remote-nrepl-ports`. 
The defcustom should be moved to the top of the file and annotated accordingly.

**Comment by aiba** [comment]
> With me it's perfectly fine to just delete everything tramp-related, as it has been nothing but a pain in the ass for me.

Yeah, same with me.  My workflow is that I always edit files locally.  When I connect to a remote repl, it's always a direct connection and I use git to sync files, rather than editing files remotely with tramp.

But I'm sympathetic to the workflow of editing files remotely with tramp.

Another strategy would be to have a master var, `cider-enable-ssh-integration` or `cider-enable-tramp-integration`, and then hide everything behind this var so non-tramp users (like us) can disable all those code paths with one var.  Would you prefer that strategy?

File: cider.el (hunk 1 above)
Comment (aiba) [reply cider.el]: I tried that, but it didn't seem like a big win, especially since this function is already entangled with an `ssh-hosts` arg.

**Comment by aiba** [comment]
>  I think it's fine for now to have two variables, both of which should be nil by default and should be well documented

OK, i'll clean this up and add docs for it.

**Comment by aiba** [comment]
OK, I think this PR is ready for another look.  I considered adding documentation to `doc/configuration.md` but I'm not sure it's worth it given how infrequently we expect people to turn this variable on, and that we might want to phase it out in the future.

**Comment by bbatsov** [comment]
I think it might actually be nice to add some section on connecting over ssh, but that's not that big of a deal.

File: CHANGELOG.md [hunk 3]
```diff
@@ -25,6 +25,7 @@
 ### Changes
 
 * [#2482](https://github.com/clojure-emacs/cider/issues/2482): Don't bind nREPL server started by `cider-jack-in` to `::` (use `localhost` instead).
+* [#1544](https://github.com/clojure-emacs/cider/issues/1544): Add a new defcustom `cider-infer-remote-nrepl-ports` to control whether we use tramp/ssh to infer remote ports.  Now defaulting to `nil` (previously it alwasy tried to infer).
```
Comment (bbatsov) [inline CHANGELOG.md]: alwasy -> alwasy

File: cider.el [hunk 4]
```diff
@@ -1327,6 +1327,11 @@ Tramp version starting 26.1 is using a `cl-defstruct' rather than vanilla VEC."
       (make-tramp-file-name :method (elt vec 0)
                             :host   (elt vec 2)))))
 
+(defcustom cider-infer-remote-nrepl-ports nil
+  "When true, cider will use ssh to try to infer nREPL ports on remote hosts."
+  :type 'boolean
+  :safe #'booleanp)
```
Comment (bbatsov) [inline cider.el]: You should also add a `:package-version` property (0.19.0).

**Comment by bbatsov** [comment]
Btw, it seems to me that the two ssh options should probably be mentioned in http://www.cider.mx/en/latest/up_and_running/, as this makes more sense than the generic configuration section.

**Comment by aiba** [comment]
> Btw, it seems to me that the two ssh options should probably be mentioned in http://www.cider.mx/en/latest/up_and_running/, as this makes more sense than the generic configuration section.

I can take a stab at docs for this.  To me it makes sense to have a section somewhere (perhaps up-and-running, or perhaps a new page) called "Connecting over SSH".  And I'm not sure I understand all of cider's behavior wrt ssh connections well enough to write this section, but I'll give it a try.

**Comment by aiba** [comment]
I think this is ready for another look.

**Pull request merged**